BXMSDOC-4559-master: Add section on custom Prometheus metrics.

sterobin opened the pull request:
@vblagoje @sutaakar , can you SME and QE review this? Valdimir said they merged it for 7.25 / 7.5 last minute and ideally would like for the new doc section to be included with the cutoff. Links below. **Please see also my questions below.**

@mramendi, can you also have a glance from the OCP perspective at the same time? So that anything you voice can be in the live conversation during SME/QE review.

Links:
* [DM doc preview](http://file.rdu.redhat.com/~sterobin/BXMSDOC-4559_DM/#prometheus-monitoring-custom-proc_execution-server)
* [PAM doc preview](http://file.rdu.redhat.com/~sterobin/BXMSDOC-4559_PAM/#prometheus-monitoring-custom-proc_execution-server) - Same, but for PAM
* [Drools doc preview](http://file.rdu.redhat.com/~sterobin/BXMSDOC-4559_DROOLS/#prometheus-monitoring-custom-proc_kie-apis) - Same, but for Drools.
* [jBPM doc preview](http://file.rdu.redhat.com/~sterobin/BXMSDOC-4559_JBPM/#prometheus-monitoring-custom-proc_kie-apis) - Same, but for jBPM.

**Questions/comments:**
* In step 1, creating the maven project and setting packaging to `jar`, doesn't the user also need to add a bunch of dependencies? I'm coincidentally also documenting a series of similar procedures for extending KIE Server capabilities based on Maciej's blogs, such as [extending KIE Server transports](http://mswiderski.blogspot.com/2015/12/kie-server-extend-kie-server-with.html), and in each case, the maven project also needs several dependencies to work properly. How about in this procedure with Prometheus?
* This procedure assumes the user has already configured Prometheus (prerequisite) either on-premise or on OCP, so the current procedure as it stands applies to OCP and on-premise alike, correct? Or is anything different for OCP in this procedure?
* The format and even section title is intentionally consistent with the other similar sections on extending KIE Server extensions that I'm currently working on. Suggestions are certainly welcome, but just FYI, if you're wondering.

vblagoje commented on doc-content/shared-kie-docs/src/main/asciidoc/KieServer/prometheus-monitoring-custom-proc.adoc:
```diff
@@ -0,0 +1,184 @@
+In this example, the `MyPrometheusMetricsProvider` class implements the `PrometheusMetricsProvider` class that you previously defined and includes your custom `ExampleCustomPrometheusMetricListener` listener class.
+--
+. To make the new metrics provider discoverable, create a `META-INF/services/org.kie.server.services.prometheus.PrometheusMetricsProvider` file in your Maven project and add the fully qualified class name of the `PrometheusMetricsProvider` implementation class within the file. For this example, the file contains the single line `org.kie.server.ext.prometheus.MyPrometheusMetricsProvider`.
+. Build your project and copy the resulting JAR file into the `~/kie-server.war/WEB-INF/lib` directory of your project.
```
@sterobin @mramendi all looks good to me except please confirm how would this step of adding customer's jar file to kie-server.war be done in Openshift.

sutaakar replied:
This can be covered as part of BAPL-1216.

sterobin replied:
@vblagoje, @sutaakar that's what I've already asked you :) I don't know, I'm just documenting according to the info you provide me, so please advise.

mramendi replied:
@vblagoje @sutaakar so custom Prometheus monitoring will require the customer to make a custom KIE server image? This sounds a tad heavy-handed to me, personally.

vblagoje commented:
@sterobin reviewed. Looks good to me. Perhaps we need a bit more clarification on how all this is to be done on Openshift. Also, if needed, we can make this Java custom Prometheus metrics project example, along with all the necessary files, available to customers? Just lmk

sutaakar commented on doc-content/shared-kie-docs/src/main/asciidoc/KieServer/prometheus-monitoring-custom-proc.adoc:
```diff
@@ -0,0 +1,184 @@
+import org.kie.server.services.api.KieContainerInstance;
+import org.optaplanner.core.impl.phase.event.PhaseLifecycleListener;
+
+public interface PrometheusMetricsProvider {
```
This is not an implementation of PrometheusMetricsProvider interface, but the interface itself.
You can find sample implementation in https://github.com/kiegroup/droolsjbpm-integration/pull/1854/files#diff-b4f2d8038e8a2130a7330adfec571191

sutaakar commented on doc-content/shared-kie-docs/src/main/asciidoc/KieServer/prometheus-monitoring-custom-proc.adoc:
```diff
@@ -0,0 +1,184 @@
+
+.Procedure
+. Create an empty Maven project and, in the `pom.xml` file for the project, set the `<packaging>` type to `jar`.
+. Implement the `org.kie.server.services.prometheus.PrometheusMetricsProvider` interface in a Java class in your project, as shown in the following example:
```
This step can be removed. The interface is available in kie-server-services-prometheus artifact.

sterobin replied:
I can remove, but question: The user doesn't declare that dependency in the prerequisite configuration unless it's on Spring Boot. All the user does is enable the `server.ext.prometheus` extension. Were users supposed to added that kie-server-services-prometheus artifact at some point but didn't? How do they then access that PrometheusMetricsProvider interface in the current documented workflow?

vblagoje replied:
@sterobin After we merge the PR PrometheusMetricsProvider will be part of the kie-server-services-prometheus artifact so users can reference it that way in their projects

sterobin replied:
@vblagoje, but again, when and where would a user have "referenced it that way in their projects"? You talking pom.xml I presume? We didn't tell them to declare that dependency in the previous configuration steps, only if they were configuring on Spring Boot. So what should we tell users to do and where? Please see the overall [Prometheus chapter](http://file.rdu.redhat.com/~sterobin/BXMSDOC-4559_DM/#prometheus-monitoring-con_execution-server) to see what I mean, and advise.

vblagoje replied:
@sterobin there is no Spring Boot involved here. They would need to reference the artifact in their pom.xml. I'll push he example project to github so you can see how exactly it would look

sutaakar commented:
I think it would be good to add also pom.xml example which customers may use. It will help them with searching for all needed dependencies and their versions (and BOM configuration and repositories containing all needed artifacts). @vblagoje What do you think about it?

sterobin commented:
@vblagoje regarding

> Perhaps we need a bit more clarification on how all this is to be done on Openshift. 

Again, that's what I've already asked you. I don't know and am just documenting according to the info you provide me. Please advise how any or all differs for OCP.

Also, @vblagoje, I agree with @sutaakar about providing the example pom with the required dependencies, etc. Please provide.

And sure, making it available to customers is fine with me, just know that if it's in github only and not in the official product distro in the Portal, then I can point to it from only the community version of the doc. Just FYI.

vblagoje commented:
@sutaakar I'll make that example now so users can reference it and have a head start. I'll push it to a github repo and then we can reference the repo in the docs and update it there if needed

sterobin commented:
@vblagoje regarding the example, just to re-iterate, we can refer to that example in github from community docs only. We will not be able to refer to it from the enterprise version later. For enterprise, examples must be in the product distro in the Portal, typically the `-source` distro file. I just want to make sure you're aware of that so you're not bummed when we don't reference it in enterprise.

sutaakar commented:
@sterobin We came to a conclusion that custom Prometheus metrics will be available just for EAP for 7.5.

sterobin commented:
@sutaakar Acknowledged. So does this mean that this PR and the content is fully QE verified for enterprise as of now? I will update the pom requirements and remove that interface step per your suggestions, and then ping you for your final verification. At that point, will this be a confirmed go as is for enterprise, or are we still thinking merge this for community only at this point and then further test this on-premise for 7.5?

sutaakar commented:
@sterobin Yes, the content will be QE verified for enterprise.

sutaakar commented:
About OCP, I think it is ok to skip it for now (no need for any specific note). Users can still use default metrics which are already available there.

sterobin commented:
@vblagoje @sutaakar I've pushed the changes. Use the same doc preview links in my description above, but **refresh the each preview that you open to get the latest.**

Summary of tweaks:
* Added example pom setup in step one, in a way that's consistent with the other "extending KIE Server" procedures that are also coming soon.
* Removed the step about adding the MetricsProvider interface, and noted that it's incorporated by the kie-server-services-prometheus dependency.
* Conditionalized two code snippets that contained jBPM-specific data and therefore not applicable to DM and Drools.
* Did NOT include the link to Vladimir's Prometheus example project because it's in his personal repo, and must be merged to the community project in order to me to be able to link to in the community docs, correct? Personal repos not permitted?

vblagoje commented:
This is it. +1 ack for merge

sutaakar commented on doc-content/shared-kie-docs/src/main/asciidoc/KieServer/prometheus-monitoring-custom-proc.adoc:
```diff
@@ -0,0 +1,316 @@
+  <dependency>
+    <groupId>org.kie.server</groupId>
+    <artifactId>kie-server-services-prometheus</artifactId>
+    <version>${kie.version}</version>
```
On some places you defined ${version.org.kie} , on other ${kie.version} .
Considering property defined above I suppose the correct version declaration is ${version.org.kie}

sterobin replied:
Oops! That was my mistake when replacing the examples. Fixed.

sutaakar commented on doc-content/shared-kie-docs/src/main/asciidoc/KieServer/prometheus-monitoring-custom-proc.adoc:
```diff
@@ -0,0 +1,316 @@
+    public PhaseLifecycleListener createPhaseLifecycleListener(String solverId) {
+        return new PhaseLifecycleListenerAdapter() {
+        };
+    }
```
createAsynchronousJobListener() and createDeploymentEventListener() has to be defined for DM too, because those methods are specified in interface PrometheusMetricsProvider

sutaakar replied:
I suppose the dependencies above need to be adjusted too, as the mentioned methods add dependency to jBPM artifacts

sterobin replied:
Fixed.

mramendi commented on doc-content/shared-kie-docs/src/main/asciidoc/KieServer/prometheus-monitoring-custom-proc.adoc:
```diff
@@ -0,0 +1,222 @@
+  </dependency>
+</dependencies>
+----
+. Implement the relevant listener from the `org.kie.server.services.prometheus.PrometheusMetricsProvider` interface as part of a custom listener class that defines your custom Prometheus metrics, as shown in the following example:
```
"as A part of THE custom listener class" would seem more correct to me

sterobin replied:
Sure, done.

mramendi commented on doc-content/shared-kie-docs/src/main/asciidoc/KieServer/prometheus-monitoring-custom-proc.adoc:
```diff
@@ -0,0 +1,222 @@
+}
+----
+
+The `PrometheusMetricsProvider` interface contains the required listeners for collecting Prometheus metrics, and is incorporated by the `kie-server-services-prometheus` dependency that you declared in your project `pom.xml` file.
```
The comma before "and" looks grammatically incorrect to me and I would split the sentence anyway for clarity. "...metrics. The interface is incorporated by..."

sterobin replied:
Done.
